### ISSUE TITLE
"root=" kernel parameter with a quoted value is not parsed correctly 

### ISSUE THREAD

**Issue by nl6720** [dracutdevs/dracut#720]
To place a space character in the `root` kernel parameter's value, the value can be quoted.
dracut does not seem to support this.
Using initramfs created with [dracut 049-4](https://www.archlinux.org/packages/extra/x86_64/dracut/) and passing `root="PARTLABEL=Arch Linux" rw resume=PARTLABEL=Swap add_efi_memmap apparmor=1 security=apparmor video=1366x768 initrd=boot\initramfs-linux.img` from the boot loader, I get:
```
Starting dracut cmdline hook...
[    2.180482] dracut: FATAL: Don't know how to handle 'root="PARTLABEL=Arch'
[    2.180605] dracut: Refusing to continue
[    2.249839] reboot: System halted
```

**Comment by nl6720** [comment]
This is still an issue with dracut 051.

**Comment by johannbg** [comment]
@nl6720 could you test if this is still an issue with the 52 release?

**Comment by nl6720** [comment]
@johannbg, yes still an issue with 52.

![VirtualBox_Arch Linux_25_02_2021_12_02_08](https://user-images.githubusercontent.com/17086862/109137063-785fa800-7761-11eb-9ead-991f1d338238.png)

**Comment by haraldh** [comment]
The linux kernel does not do this.

**Comment by haraldh** [comment]
best to replace the space with s.th. else like `_`

**Comment by haraldh** [comment]
To elaborate: The linux kernel does not honor `"`, so even if dracut would do that, you could accidentally trigger some other valid kernel option after the space. So dracut better follows the model the linux kernel uses to parse the kernel command line.

**Comment by nl6720** [comment]
I disagree with your reasoning. dracut (presumably by using udev) supports more `root=` formats (`root=LABEL=`, `root=UUID=`, etc.) than the kernel, so I don't see why it couldn't support quoted spaces.

`root="PARTLABEL=Arch Linux"` works fine with [mkinitcpio](/archlinux/mkinitcpio/), so I consider this a bug in dracut.

**Comment by haraldh** [comment]
As I said, this will conflict with the kernel parsing.
E.g. `PARTLABEL="Arch debug Partition"`

and kernel command line: `root="PARTLABEL=Arch debug Partition"`

will give you a lot of debug output, because the kernel parses:
```
root="PARTLABEL=Arch
debug
Partition"
```
as parameters.

I really don't want to see other scenarios.

**Comment by nl6720** [comment]
Why does it matter if it conflicts with the kernel parsing? dracut could parse the parameter itself and convert it to a format the kernel understands.

**Comment by haraldh** [comment]
The kernel parses the kernel command line on its own, before the initramfs is even loaded and executed.

**Comment by haraldh** [comment]
To elaborate, in my example above, you would turn on the kernel "debug" parameter, which significantly affects your system.

**Comment by LMBernardo** [comment]
The 4.x kernels support double-quoting parameters with spaces, at least according to the [documentation](https://www.kernel.org/doc/html/v4.10/admin-guide/kernel-parameters.html). I doubt they would remove this in 5.x, and the relevant functions appear to just use strcmp which doesn't care about spaces, though it's not explicitly mentioned in this 5.x docs anymore. Considering mkinitcpio allows double-quoting kernel parameters with spaces, I would at least consider this to be an important feature request if not a bug.

**Comment by haraldh** [comment]
I stand corrected and will update all parameter parsing functions.

**Comment by haraldh** [comment]
ok, WIP: https://github.com/dracutdevs/dracut/pull/1143

@nl6720 : `mkinitcpio` might fail, if you have two consecutive spaces in the `PARTLABEL` or in any kernel parameter, looking at https://github.com/archlinux/mkinitcpio/blob/master/init_functions#L209

**Comment by nl6720** [comment]
@grazzolini

>  `mkinitcpio` might fail, if you have two consecutive spaces in the `PARTLABEL` or in any kernel parameter, looking at https://github.com/archlinux/mkinitcpio/blob/master/init_functions#L209

**Comment by grazzolini** [comment]
@nl6720 indeed, it might fail. This is a weird one. Can you open an issue on mkinitcpio as well?

**Comment by haraldh** [comment]
Should be fixed with #1143
